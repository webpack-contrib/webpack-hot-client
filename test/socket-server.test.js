const http = require('http');

const WebSocket = require('ws');

const { getServer, payload } = require('../lib/socket-server');
const getOptions = require('../lib/options');

const options = getOptions({ logLevel: 'silent' });

const createServer = (port, host, callback) =>
  new Promise((resolve) => {
    const server = http.createServer();
    server.on('close', () => {
      resolve();
    });
    server.listen(port, host, callback.bind(null, server));
  });

const getSocket = (port, host) => new WebSocket(`ws://${host}:${port}`);

describe('socket server', () => {
  test('getServer', (done) => {
    const server = getServer(options);
    const { broadcast, close, send } = server;

    expect(broadcast).toBeDefined();
    expect(close).toBeDefined();
    expect(send).toBeDefined();

    server.on('listening', () => {
      const { host, port } = server;
      expect(host).toBe('127.0.0.1');
      expect(port).toBeGreaterThan(0);

      close(done);
    });
  });

  test('getServer: { server }', () =>
    createServer(1337, '127.0.0.1', (server) => {
      const opts = getOptions({ server });
      const { close, host, port } = getServer(opts);

      expect(host).toBe('127.0.0.1');
      expect(port).toBe(1337);

      server.close(close);
    }));

  test('payload', () => {
    expect(payload('test', { batman: 'superman ' })).toMatchSnapshot();
  });

  test('broadcast', (done) => {
    const server = getServer(options);
    const { broadcast, close } = server;

    server.on('listening', () => {
      const { host, port } = server;
      const catcher = getSocket(port, host);

      catcher.on('message', (data) => {
        expect(data).toMatchSnapshot();
        close(done);
      });

      catcher.on('open', () => {
        broadcast(payload('broadcast', { received: true }));
      });
    });
  });

  test('socket broadcast', (done) => {
    const server = getServer(options);
    const { close } = server;

    server.on('listening', () => {
      const { host, port } = server;
      const pitcher = getSocket(port, host);
      const catcher = getSocket(port, host);

      catcher.on('message', (data) => {
        expect(data).toMatchSnapshot();
        close(done);
      });

      pitcher.on('open', () => {
        pitcher.send(payload('broadcast', { received: true }));
      });
    });
  });

  test('send via stats', (done) => {
    const stats = {
      context: process.cwd(),
      toJson: () => {
        return { hash: '111111', warnings: [] };
      },
    };
    const opts = getOptions({ logLevel: 'silent', stats });
    const server = getServer(opts);
    const { close } = server;

    server.on('listening', () => {
      const { host, port } = server;
      const catcher = getSocket(port, host);

      catcher.on('message', (raw) => {
        const data = JSON.parse(raw);
        if (data.type === 'ok') {
          expect(data).toMatchSnapshot();
          close(done);
        }
      });
    });
  });

  test('errors', (done) => {
    const server = getServer(options);
    const { close } = server;

    server.on('listening', () => {
      const { host, port, send } = server;
      const catcher = getSocket(port, host);

      catcher.on('message', (raw) => {
        const data = JSON.parse(raw);
        if (data.type === 'errors') {
          expect(data).toMatchSnapshot();
          close(done);
        }
      });

      catcher.on('open', () => {
        send({
          errors: ['test error'],
        });
      });
    });
  });

  test('hash-ok', (done) => {
    const server = getServer(options);
    const { close } = server;

    server.on('listening', () => {
      const { host, port, send } = server;
      const catcher = getSocket(port, host);

      catcher.on('message', (raw) => {
        const data = JSON.parse(raw);
        if (data.type === 'ok') {
          expect(data).toMatchSnapshot();
          close(done);
        }
      });

      catcher.on('open', () => {
        send({
          hash: '000000',
          warnings: [],
        });
      });
    });
  });

  test('no-change', (done) => {
    const server = getServer(options);
    const { close } = server;

    server.on('listening', () => {
      const { host, port, send } = server;
      const catcher = getSocket(port, host);

      catcher.on('message', (data) => {
        expect(data).toMatchSnapshot();
        close(done);
      });

      catcher.on('open', () => {
        send({
          assets: [{ emitted: false }],
        });
      });
    });
  });

  test('hash-ok', (done) => {
    const server = getServer(options);
    const { close } = server;

    server.on('listening', () => {
      const { host, port, send } = server;
      const catcher = getSocket(port, host);

      catcher.on('message', (raw) => {
        const data = JSON.parse(raw);
        if (data.type === 'warnings') {
          expect(data).toMatchSnapshot();
          close(done);
        }
      });

      catcher.on('open', () => {
        send({
          hash: '000000',
          warnings: ['test warning'],
        });
      });
    });
  });
});
