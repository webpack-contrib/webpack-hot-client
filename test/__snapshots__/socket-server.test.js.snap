// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`socket server broadcast 1`] = `"{\\"type\\":\\"broadcast\\",\\"data\\":{\\"received\\":true}}"`;

exports[`socket server errors 1`] = `
Object {
  "data": Object {
    "errors": Array [
      "test error",
    ],
  },
  "type": "errors",
}
`;

exports[`socket server hash-ok 1`] = `
Object {
  "data": Object {
    "hash": "000000",
  },
  "type": "ok",
}
`;

exports[`socket server hash-ok 2`] = `
Object {
  "data": Object {
    "warnings": Array [
      "test warning",
    ],
  },
  "type": "warnings",
}
`;

exports[`socket server no-change 1`] = `"{\\"type\\":\\"no-change\\"}"`;

exports[`socket server payload 1`] = `"{\\"type\\":\\"test\\",\\"data\\":{\\"batman\\":\\"superman \\"}}"`;

exports[`socket server send via stats 1`] = `
Object {
  "data": Object {
    "hash": "111111",
  },
  "type": "ok",
}
`;

exports[`socket server socket broadcast 1`] = `"{\\"received\\":true}"`;
